### ISSUE TITLE
code-completion doesn't use next suggestion on pressing tab multiple times

### ISSUE THREAD

**Issue by DomiR** [Microsoft/vscode#28215]
**TL;DR** the <kbd>tab</kbd> key should insert first suggestion, but then cycle through suggestions, replacing the last completion.

Sublime autocompletion can be configured so that when pressing the <kbd>tab</kbd> multiple times it will cycle through the autocompletions by inserting the first suggestion (closing the suggestion box), then replacing the suggestion by succeeding ones when pressing <kbd>tab</kbd> repeatedly.

Problem description: 
Visually:
```
// e.g. 1. typing`whatI` will render both options in suggestion box
whatI|
      -whatIGet   // => 2. first option will be autocompleted on `tab` closing the suggestion box
      -whatIWant // => 3. on second `tab` this should replace the first (wrong) suggestion

function whatIWant() {}

funciton whatIGet() {}
``` 

Textually:
1. Type whatI
2. Suggestions will pop up
3. Wrong suggestion is highlighted
4. <kbd>tab</kbd> to complete wrong suggestion
5. Suggestions-popup disappears
6. Still at the end of inserted suggestion press <kbd>tab</kbd> again
7. Wrong suggestion should be replaced by next suggestion in list

Currently the corresponding controllers don't seem to support this, as accepting the current suggestion to my understanding will clear the suggestion list.

https://github.com/Microsoft/vscode/blob/3215f19074d49e32d5ab55203923865910d4d5ca/src/vs/editor/contrib/suggest/browser/suggestController.ts
https://github.com/Microsoft/vscode/blob/794b435b27bcf4a92cc0a6345b97821e3c827dcc/src/vs/editor/contrib/suggest/browser/suggestWidget.ts

How are people getting around having to press arrow-down and arrow-up :man_facepalming:? I somehow never get the correct suggestion at the top and I quite frankly don't have the most flexible pinky 😢

**Comment by andreyorst** [comment]
> 4. <kbd>tab</kbd> to complete wrong suggestion
> 5. Suggestions-popup disappears

Yeah, but why close popup on <kbd>tab</kbd>?

If your _correct_ suggestion isn't next in the list, and popup is closed you can run into a situation when you rapidly pressed tab key and gone too far in your list, then you could use <kbd>↑</kbd> to go up, but there is no popup so you wil probably change line. 
Yes, I know that you can use <kbd>shift</kbd>+<kbd>tab</kbd> to go backwards in the list, but I think it is better to see list, because it is changing all the time, and you dont need to remember how far your suggestion is.

Well maybe someone dont like the fact that when completion is done, no matter wrong or right one, the list is still there and hovers above the code, hiding it from eyesight, and closing popup after first <kbd>tab</kbd> is good for them, but there are other people like me who like the opposite behavior, when the popup is always open until you close it by typing.

**Comment by Naramsim** [comment]
[![enter image description here][1]][1]

A nice GIF showing what we can do in Sublime

  [1]: https://i.stack.imgur.com/hSCWY.gif

https://stackoverflow.com/questions/48883721/cycle-to-previous-suggestions-in-vscode

**Comment by ryanflorence** [comment]
This isn't exactly a duplicate of #34694, supertab in vim keeps the popup open, and doesn't pop up until you hit tab.

![screenflow](https://user-images.githubusercontent.com/100200/38039849-041b7b54-3263-11e8-8381-3594f3dc076e.gif)

Code completion the way VSCode handles it is really disruptive and slower than just typing out what I want :\

**Comment by andreyorst** [comment]
@ryanflorence well, the main idea of any modern autocompletion system is fuzziness. Thats is, you don't need to scroll through list most of the time. Let me demonstrate this:
![fuzzy](https://user-images.githubusercontent.com/19470159/38047764-25a2a4fa-32cc-11e8-82f9-1c1e8fdabdc7.gif)
At first I typed `sii` wich gave me **s**ome_funct**i**on_w**i**th_name_that_is_not_shorter()
And then I type `sul` and got **s**ome_f**u**nction__that_has_unnecessary_**l**ong_name()

However it is not always possible to match with fuzzy, like here:
![image](https://user-images.githubusercontent.com/19470159/38048123-38f68322-32cd-11e8-89d0-4711b13a88d6.png)
(Imagine that I need that int variable)
and therefore I likely need to spam a tab key some times (I actually use vim with deoplete, so I can), but again, usually I don't, because like here, it is faster to hit <kbd>c</kbd> twice.

BTW regarding to topivc, why supertab? It's not nearly as good as standard vim completion system, and far behind other async plugins like deoplete or neovim-completion-manager, or even that bloated YCM

**Comment by davidnorthetal** [comment]
@andreyorst fuzzy searches are not orthogonal to cycling and I'm quite sure everyone here understands how fuzzy searches work

you fuzzy-search when you know the symbol you're looking for (or have a strong suspicion) but don't want to write it all out yourself
you cycle when you want to explore your options

**Comment by andreyorst** [comment]
> so let's stay ontopic

ok, you want cycling. I'm not against cycling either. Let me explain myself. I'm against programmers who cycle around after every keypress, browsing kilometers of completion items, just to get single one, wich they even know about, but just because it's so fun to cycle around they will do it. In most cases you need to press one extra key (rarely two) to fuzzy match through hell lot of items in the list, but they will hold <kbd>Tab</kbd> and wait until chooser will reach their desired completion item. Thats why vscode completion system is good. It disciplines programmer to use his brain instead of brute force. It is hard to figure out how much more productive you will be if you will not scroll every time, and fuzzy match through. 

However cycling may benefit in some cases, when you have to deal with messed codebase, with large classes with tons of fields and methods, and you just want to examine it. The problem is only the fact that arrows used for that.

**Comment by jrieken** [comment]
fyi - I have an implementation this in https://github.com/Microsoft/vscode/pull/58871. It will be behind a setting and disabled by default. Now a remaining question: How should this be called? A need a good name for the setting. Anyone having ideas? @DomiR how does sublime call this?

**Comment by jrieken** [comment]
![sep-18-2018 10-45-51](https://user-images.githubusercontent.com/1794099/45675639-103c1f00-bb30-11e8-9623-f010ed63a40f.gif)

This is how it works

* it is only enabled when the suggestion has been accepted with `tab`
* it won't work when the accepted suggestion is a snippets
* when cycling through the alternatives, snippets-suggestions are skipped
* only narrowed-down list of suggestions is cycled through (e.g. console.g -> group, groupCollapsed, groupEnd)

**Comment by svipben** [comment]
@jrieken amazing work!

**Comment by DomiR** [comment]
😍 this is perfection. I loved this feature in Sublime and it would be great to get this into VSCode. You are the best, thank you so much!

They call it "tab_completion", see here: https://www.sublimetext.com/docs/2/tab_completion.html

**Comment by jrieken** [comment]
Hm, we have `editor.tabCompletion` but that currently on expands snippets and only when the cursor is after a snippet prefix. We can extend that option to be `off | onlySnippets | on` where todays `true` maps onto `onlySnippets`.

**Comment by jrieken** [comment]
This is now in master and will be in tomorrows Insiders build: [`http://code.visualstudio.com/insiders/`](http://code.visualstudio.com/insiders/)!

Try it and let me know what you think. This is how it works

* the `editor.tabCompletion` setting has now new values `off` (was `false`), `onlySnippets` (was `true`), and `on` for the full tab completion behaviour 
* when tab completion is `on` it will expand a prefix (when the suggest widget isn't showing)
* when tab completion is `on` and no suggestion could be found it will insert `\t`
* when tab completion is `on` it will cycle through other suggestions given the current suggestion has been accepted via <kbd>tab</kbd> or was tab-completed

**Comment by svipben** [comment]
@jrieken What can I say... It's the best feature, I'm in heaven 😄

I'm using these settings and it feels amazing, no more annoying suggestions popup and I can easily autocomplete my changes by pressing `tab`.


```
  "editor.quickSuggestions": false,
  "editor.suggestOnTriggerCharacters": false,
  "editor.tabCompletion": "on",
  "emmet.triggerExpansionOnTab": true,
```

You did really great job because it surpassed my expactantions, it works perfectly!!!

**Comment by jrieken** [comment]
Thanks for the nice feedback!

**Comment by salmankhann** [comment]
since the conversation is happening here, I thought of combining my feedback

@jrieken although the current feature in the Insiders is better than before, I was looking for something a bit different that is implement in Xcode. 

See how tab keeps the IntelliSense open, as well as keeps highlighting matched elements.


![443dkek1qmoxy6uulp](https://user-images.githubusercontent.com/2960078/46394594-a0ff2700-c6fa-11e8-9cc3-1df05b84e2fa.gif)

**Comment by PEZ** [comment]
Is there going to be a [when clause context](https://code.visualstudio.com/docs/getstarted/keybindings#_when-clause-contexts) that can detect when the suggestions are in this cycle mode? I am the author of an extension where the `tab` key reformats enclosing code and that makes this new feature unavailable for my users if they keep the key binding.

Maybe @jrieken knows the answer to my question?

**Comment by svipben** [comment]
Guys, this issue was closed, if you want to ask a feature request or something like that, please open a new issue.
